File: aat-lib/src/main/java/ch/bailu/aat_lib/preferences/location/SolidAltitude.java
package ch.bailu.aat_lib.preferences.location;

import ch.bailu.aat_lib.logger.AppLog;
import ch.bailu.aat_lib.preferences.SolidInteger;
import ch.bailu.aat_lib.preferences.StorageInterface;
import ch.bailu.aat_lib.preferences.general.SolidUnit;

public class SolidAltitude extends SolidInteger {

    private final int unit;


    public SolidAltitude(StorageInterface s, String k, int u) {
        super(s, k);
        unit = u;
    }


    public String addUnit(String s) {
        return s + " [" + SolidUnit.ALT_UNIT[unit] + "]";
    }


    @Override
    public String getValueAsString() {
        return String.valueOf(Math.round(getValue() * SolidUnit.ALT_FACTOR[unit]));
    }


    @Override
    public void setValueFromString(String s) {
        try {
            setValue(Math.round(Float.parseFloat(s) / SolidUnit.ALT_FACTOR[unit]));
        } catch (NumberFormatException e) {
            AppLog.e(this, e);
        }
    }


}

File: aat-lib/src/main/java/ch/bailu/aat_lib/preferences/location/SolidAltitude.kt
package ch.bailu.aat_lib.preferences.location

import ch.bailu.aat_lib.exception.ValidationException
import ch.bailu.aat_lib.logger.AppLog.e
import ch.bailu.aat_lib.preferences.SolidInteger
import ch.bailu.aat_lib.preferences.StorageInterface
import ch.bailu.aat_lib.preferences.general.SolidUnit
import ch.bailu.aat_lib.resources.Res
import kotlin.math.roundToInt

open class SolidAltitude(storage: StorageInterface, key: String, private val unit: Int) : SolidInteger(storage, key) {
    fun addUnit(s: String): String {
        return s + " [" + SolidUnit.ALT_UNIT[unit] + "]"
    }

    override fun getValueAsString(): String {
        return (getValue() * SolidUnit.ALT_FACTOR[unit]).roundToInt().toString()
    }

    override fun setValueFromString(string: String) {
        if (!validate(string)) {
            throw ValidationException(Res.str().error_integer())
        }

        try {
            setValue((string.toFloat() / SolidUnit.ALT_FACTOR[unit]).roundToInt())
        } catch (e: NumberFormatException) {
            e(this, e)
        }
    }
}
